### ISSUE TITLE
Spec interface for build information (like lib names, includes, etc.)

### ISSUE THREAD

**Issue by tgamblin** [LLNL/spack#1821]
Per discussion in #1682, reposting this for discussion.

One thing I thought about when looking at this PR: I don't _really_ like it that some dependents still search for their dependencies libraries, though I realize that it can be necessary.  Basically there are two guiding principles here:
1. The dependencies know themselves best, and if they can expose their build information consistently, they should do that and let dependents access it.
2. Dependents should prefer to ask dependencies for build information (lib locations, etc.) and only search if necessary.  Searching complicates build logic and Spack should already know where things are because it already installed the dependencies.

(2) breaks down for external packages, which Spack didn't install and which may even be laid out differently than Spack would lay them out (e.g. a install of boost might have al the weird lib suffixes that boost supports, and I think Spack wouldn't currently know what to do with that).  Even better is that `find_libraries` is hidden behind a property interface so the client code doesn't know how the library names are being assigned.

This PR follows both of these principles for `LAPACK` and `BLAS`.  Using properties on specs for things like `blas_libs` and `lapack_libs` is great.  It doesn't follow these for `fftw`, `arpack`, `superlu`, etc.  See the `install` method for `armadillo`, which does its own `find` routines.

I think this PR makes it clearer to me what a good convention for passing args among packages in a build should look like.  You want to ask the package for a property, and the package needs to know a) the property name and b) the `vdep` it's being asked as.  For non-virtual packages, (b) doesn't matter and is always just the plain package name, but when you're asked to provide the libs/headers for a particular _interface_, you need to know the interface name. 

I guess the question is whether this should be a convention or a more formal interface.  The convention currently looks like:

``` python
    spec['blas'].blas_libs
    spec['lapack'].lapack_libs
```

Slightly annoying things about it:
1. You have to write the interface twice even though you know what you asked for.
1. The interface name is written once as a string and once as a property name prefix.

So, think we could have a convention like this:

``` python

def provider_property(function):
    """Function object for grabbing build properties from dependency Packages."""
    def __init__(self, name):
        self.name = name

    def __call__(self, provider):
        """Get provider-specific property if available, or default to generic property."""
        specific = provider.replace('-', '_') + '_' + self.name
        dep = spec[provider]
        if hasattr(dep, specific):
            return getattr(dep, specific)
        else:
            return getattr(dep, self.name)  # default to generic attribute


class Spec(object):
    ...
    libs = provider_property('libs')
    includes = provider_property('includes')
```

Now the interface looks like this:

``` python
    spec.libs('blas')
    spec.libs('lapack')
```

And you've got a convention for properties to add to packages (`libs`, `<provider>_libs`, etc.).  `Package` could provide a default implementation:

``` python
class Package(object):
    ...
    def libs(self):
        # requiring installation allows us to be more generic b/c we can search,
        # but it does restrict when you can call the method.
        assert(self.installed)    

        return find_libraries(... something that returns all the libs in `lib` or `lib64`...)
```

And packages that need to can add things like `blas_libs`, `lapack_libs`, `superlu_dist_libs`, etc. as they do in this PR.  Packages like `arpack` and `superlu` can do nothing b/c we find their libs automatically and they're available through, e.g. `spec.libs('arpack')`.

I guess the question is whether the interface above is so much better than the simple convention we've got that it's worthwhile.  The current convention is not so bad.  I think it boils down which is simpler to document and teach to package authors.  

The other question is whether this convention is worthwhile for non-virtuals.  Sometimes the client needs to be specific about which lib names it needs, in which case find_libraries seems like the right thing to do.

What do you all think?

**Comment by tgamblin** [comment]
Response from @davydden: 

> 1. You have to write the interface twice even though you know what you asked for.

one have to do that because a single package can provide different virtual packages (e.g. mkl provides blas and lapack and scalapack). Would the convention you suggested work here as well?

> The other question is whether this convention is worthwhile for non-virtuals. Sometimes the client needs to be specific about which lib names it needs, in which case find_libraries seems like the right thing to do.

i would say for most packages the library names are simple (`libarpack`, `libsuperlu_dist`, etc) and for simple configure systems `-L%s -larpack' % spec['arpac'].prefix.lib` will do. For complicated cases like `trilinos` and `oce`, which ave tens (!) of libraries, dependents (i.e. `dealii`) know which libs to pick up from a prefix. So using this class is an overkill and hard to maintain as the list of libraries may change between versions.

So i would not adopt this approach to every package, but only to a few virtuals (`fftw` is probably the only one missing).


What should be systematically addressed is `lib` folder. I don't have a good overview, but i guess libraires may end up both in `prefix/lib` and `prefix/lib64` on 64bit systems. In some packages `%s/lib` is hardcoded, in others it is used as `spec[package].prefix.lib`, but I am not sure that each package somehow checks whether or not things are installed in `lib` or `lib64`. If that's the case already, then `spec[package].prefix.lib` should be used throughout.

**Comment by tgamblin** [comment]
> one have to do that because a single package can provide different virtual packages (e.g. mkl provides blas and lapack and scalapack). Would the convention you suggested work here as well?

Yes.  The idea is that dependents write `spec.libs('blas')`, and that ends up getting `blas_libs` from `spec['blas']`, and falls back to `spec['bias'].libs` if `spec['blas'].blas_libs` doesn't exist.  So it's simpler than client code.  I am starting to think that the improvement is too small to justify the complicated mechanism.  I tend to agree with you that this makes the most sense for a few virtuals, in which case `spec['blas'].blas_libs` is probably clear enough.  Also it's clearer what is going on if you read that vs. `spec.libs('blas')`.  So this change is probably not necessary.

RE: `lib` vs. `lib64`: We currently don't document or enforce a convention about this, but in Spack we could probably just force packages to use `lib`, since each build has only a single architecture anyway.  The current policy is that packages do whatever they do by default (I believe most of them use `lib`) and we install that.  The compilers are smart enough to check for both `lib` and `lib64` directories, and they'll add the one that exists.  I do not know of any cases in Spack where both exist, but I haven't dug deeply into this.  Only one ever really should exist b/c, again, builds are single-arch.

**Comment by tgamblin** [comment]
Ok after looking at #1823, where @alalazo added a `__getattr__` to `spec` to delegate certain calls to package, I think we need to at least document this convention.  @alalazo do you have any thoughts on this?

**Comment by alalazo** [comment]
> Basically there are two guiding principles here:
> ...
> Even better is that `find_libraries` is hidden behind a property interface so the client code doesn't know how the library names are being assigned.

I tried to design things that way for `blas` and `lapack` providers, so I am really glad that this is the impression the code gives!

> I think this PR makes it clearer to me what a good convention for passing args among packages in a build should look like. You want to ask the package for a property, and the package needs to know a) the property name and b) the vdep it's being asked as. For non-virtual packages, (b) doesn't matter and is always just the plain package name, but when you're asked to provide the libs/headers for a particular interface, you need to know the interface name.
> ...
> Slightly annoying things about it:
> 1. You have to write the interface twice even though you know what you asked for.
> 1. The interface name is written once as a string and once as a property name prefix.
> ...
> I guess the question is whether the interface above is so much better than the simple convention we've got that it's worthwhile. The current convention is not so bad. I think it boils down which is simpler to document and teach to package authors.
>
>The other question is whether this convention is worthwhile for non-virtuals. Sometimes the client needs to be specific about which lib names it needs, in which case find_libraries seems like the right thing to do.

I think it's worthwhile to extend the convention to non-virtuals, as it may result in a bit more concise and intuitive syntax for package writers. That was also the sense of #1186. The syntax you propose above is fine with me. Another possibility that could be explored is :
```python
spec['blas'].libs
spec['lapack'].libs
```
Something like this could be done reusing a lot of what you propose, with the exception that `Spec.__getitem__(self, item)` should return a `SpecProxy` that could record the `item` string and use it in a few forwarding methods like `libs`.

> I am thinking there should really be an accessor function that grabs these package attributes from a dependency, like the libs function I outlined above. Basically I think there should be an explicit convention rather than using __getattr__, because I think using __getattr__ on any missing attribute is going to be really confusing in the long run. There are cases where delegating to the package makes sense, but I think in general people should keep some mental separation between the spec and the package.

If we are going to extend the convention I agree we should be more explicit. Better :  as soon as we agree on an interface and a logic that works for most (if not all) the cases we have I am vailable to implement it :smile:

**Comment by alalazo** [comment]
>We currently don't document or enforce a convention about this, but in Spack we could probably just force packages to use lib, since each build has only a single architecture anyway. The current policy is that packages do whatever they do by default (I believe most of them use lib) and we install that. The compilers are smart enough to check for both lib and lib64 directories, and they'll add the one that exists. I do not know of any cases in Spack where both exist, but I haven't dug deeply into this. Only one ever really should exist b/c, again, builds are single-arch.

A relevant exception that comes to my mind are Intel related packages : they usually break any convention on folder names...

**Comment by davydden** [comment]
> I think it's worthwhile to extend the convention to non-virtuals

As long as packages are not forced to provide list of libraries, this is ok. That is, default implementation in `Package` class should throw an error if someone tries to use `spec[package].libs`. For a record, installation of Trilinos has 276 (!) libraries inside `lib`. 

> they usually break any convention on folder names...

oh yes... :-( that's why MKL package currently symlinks things to comply with `prefix.lib`.

**Comment by adamjstewart** [comment]
@tgamblin Can this be closed now that #1875 is merged?

**Comment by tgamblin** [comment]
Yep!
